src/codegen.cpp

```diff
@@ -4023,6 +4028,7 @@ static jl_cgval_t emit_invoke(jl_codectx_t &ctx, const jl_cgval_t &lival, const
                     if (ctx.use_cache) {
                         // optimization: emit the correct name immediately, if we know it
                         // TODO: use `emitted` map here too to try to consolidate names?
+                        // WARNING: isspecsig is protected by the codegen-lock. If that lock is removed, then the isspecsig load needs to be properly atomically sequenced with this.
```

Comment:
@pchintalapudi do you think we can just move the isspecsig code before the setting of the specptr, then use acquire loads here? since you have been working on removing the codegen-lock (and here I am just adding a comment about the current code)

Reply:
I think that could be part of the solution, but I suspect that some of this code needs to be reengineered (e.g. in jitlayers when setting the invokeptr, 1. we should first cmpxchg the specptr against null, 2. if success then set specsig/invokeptr, 3. if fail busyloop on invokeptr until it's set and proceed, so that we cannot get a mismatched specsig, invokeptr, and specptr).

Reply:
Yeah, there probably needs to be some sort of ordering there like that. I think that ordering makes sense. Thus we can choose whether to do relaxed loads for both, or cmpswap acq-rel of specptr followed by specsig followed by invokeptr (operating somewhat like a write-once sequence lock), or an acquire load invokeptr followed by relaxed reads of the others (if successful).

We may also want to attempt to prevent multiple threads from doing compilation simultaneously for the same CodeInfo via an out-of-band shared table (in jitlayers) of all in-progress work (currently it only starts to track "in flight" objects after codegen finishes)

There is also a writelock in the Method object, but that might be too course-grained for this (we intended only to use it for quick things like adding new backedges or new specializations).

> if fail busyloop on invokeptr until it's set and proceed

cmpswap does not fail spurious, either it updates the pointer (because it was NULL), or it returns a non-NULL value (which we will never replace again, since it would be mismatched during the transition and thus we cannot do that atomically)